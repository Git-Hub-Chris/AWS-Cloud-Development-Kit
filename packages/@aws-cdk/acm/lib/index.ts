export * from './certificate';
export * from './certificate-ref';

// The L1 library for AWS::CertificateManager:
export * from '../cfn/certificatemanager';
