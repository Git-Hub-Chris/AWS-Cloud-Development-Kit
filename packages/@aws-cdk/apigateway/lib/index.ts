// The L1 Library for AWS::ApiGateway:
export * from '../cfn/apigateway';
