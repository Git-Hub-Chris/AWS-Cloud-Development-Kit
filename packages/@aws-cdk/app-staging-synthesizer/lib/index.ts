export * from './default-staging-stack';
export * from './synthesizer';
