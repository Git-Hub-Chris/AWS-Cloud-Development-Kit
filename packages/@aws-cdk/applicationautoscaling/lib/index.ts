// The L1 Library for AWS::ApplicationAutoScaling:
export * from '../cfn/applicationautoscaling';
