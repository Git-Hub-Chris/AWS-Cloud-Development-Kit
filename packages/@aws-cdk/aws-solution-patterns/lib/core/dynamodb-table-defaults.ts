import dynamodb = require('@aws-cdk/aws-dynamodb');

const DefaultTableProps: dynamodb.TableProps | any = {
  partitionKey: {
    name: 'id',
    type: dynamodb.AttributeType.STRING
  }
};

export { DefaultTableProps };