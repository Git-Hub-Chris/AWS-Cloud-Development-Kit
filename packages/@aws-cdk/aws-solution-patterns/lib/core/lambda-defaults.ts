import lambda = require('@aws-cdk/aws-lambda');
import logs = require('@aws-cdk/aws-logs');

const DefaultFunctionProps: lambda.FunctionProps | any = {
  logRetention: logs.RetentionDays.THREE_MONTHS,
  environment: {
    LOGLEVEL: 'INFO'
  }
};

export { DefaultFunctionProps };