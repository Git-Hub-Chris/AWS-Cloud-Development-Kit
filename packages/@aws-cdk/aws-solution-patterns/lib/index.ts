export * from './lambda-dynamodb/index';