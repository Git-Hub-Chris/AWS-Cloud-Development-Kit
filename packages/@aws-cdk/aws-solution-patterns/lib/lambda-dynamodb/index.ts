import dynamodb = require('@aws-cdk/aws-dynamodb');
import lambda = require('@aws-cdk/aws-lambda');
import { Construct } from '@aws-cdk/core';
import { DefaultTableProps } from '../core/dynamodb-table-defaults';
import { DefaultFunctionProps } from '../core/lambda-defaults';
import { overrideProps } from '../core/utils';

export interface LambdaToDynamoDBProps {
  /**
   * Whether to create a new lambda function or use an existing lambda function
   *
   * If set to false, you must provide a lambda function object as `existingObj`
   *
   * @default true
   */
  readonly deployLambda: boolean,
  /**
   * [disable-awslint:ref-via-interface]
   * Existing instance of Lambda Function object
   *
   * If `deploy` is set to false only then this property is required
   *
   * @default - None
   */
  readonly existingLambdaObj?: lambda.Function,
  /**
   * Optional user provided props to override the default props
   *
   * If `deploy` is set to true only then this property is required
   *
   * @default - Default props are used
   */
  readonly lambdaFunctionProps?: lambda.FunctionProps,
  /**
   * Optional user provided props to override the default props
   *
   * @default - Default props are used
   */
  readonly dynamoTableProps?: dynamodb.TableProps
}

export class LambdaToDynamoDB extends Construct {
  private fn: lambda.Function;
  private table: dynamodb.Table;

  /**
   * @param {cdk.App} scope
   * @param {string} id
   * @param {cdk.StackProps=} props
   */
  constructor(scope: Construct, id: string, props: LambdaToDynamoDBProps) {
    super(scope, id);

    // Set the default props for DynamoDB table
    if (props.dynamoTableProps) {
      const dynamoTableProps = overrideProps(DefaultTableProps, props.dynamoTableProps);
      this.table = new dynamodb.Table(this, 'DynamoTable', dynamoTableProps);
    } else {
      this.table = new dynamodb.Table(this, 'DynamoTable', DefaultTableProps);
    }

    // Conditional lambda function creation
    // If deployLambda == false
    if (props.hasOwnProperty('deployLambda') && props.deployLambda === false) {
      if (props.existingLambdaObj) {
        this.fn = props.existingLambdaObj;
      } else {
        throw Error('Missing existingLambdaObj from props for deploy = false');
      }
    // If deployLambda == true
    } else {
      if (props.lambdaFunctionProps) {
        const lambdaFunctionProps = overrideProps(DefaultFunctionProps, props.lambdaFunctionProps);
        this.fn = new lambda.Function(this, 'LambdaFunction', lambdaFunctionProps);
      } else {
        throw Error('Missing lambdaFunctionProps from props for deploy = true');
      }
    }

    this.fn.addEnvironment('DDB_TABLE_NAME', this.table.tableName);

    this.table.grantReadWriteData(this.fn.grantPrincipal);
  }

  public dynamoTable(): dynamodb.Table {
    return this.table;
  }

  public lambdaFunction(): lambda.Function {
    return this.fn;
  }
}