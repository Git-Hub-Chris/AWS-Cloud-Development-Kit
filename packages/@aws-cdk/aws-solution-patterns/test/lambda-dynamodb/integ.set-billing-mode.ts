/// !cdk-integ *
import dynamodb = require('@aws-cdk/aws-dynamodb');
import lambda = require('@aws-cdk/aws-lambda');
import { App, Stack } from "@aws-cdk/core";
import path = require('path');
import { LambdaToDynamoDB } from "../../lib";
const app = new App();

// Change the billing mode to PAY_PER_REQUEST
const stack = new Stack(app, 'test-lambda-dynamodb-stack');

new LambdaToDynamoDB(stack, 'test-lambda-dynamodb-stack', {
    dynamoTableProps: {
        billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
        partitionKey: {
            name: 'id',
            type: dynamodb.AttributeType.STRING
        }
    },
    deployLambda: true,
    lambdaFunctionProps: {
        code: lambda.Code.fromAsset(path.join(__dirname, 'lambda/handler.zip')),
        runtime: lambda.Runtime.NODEJS_10_X,
        handler: 'index.handler',
    }
});

app.synth();