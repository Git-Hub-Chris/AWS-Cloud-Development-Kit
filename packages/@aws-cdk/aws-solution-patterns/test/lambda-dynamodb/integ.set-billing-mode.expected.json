{
  "Resources": {
    "testlambdadynamodbstackDynamoTable8138E93B": {
      "Type": "AWS::DynamoDB::Table",
      "Properties": {
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH"
          }
        ],
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S"
          }
        ],
        "BillingMode": "PAY_PER_REQUEST"
      },
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "testlambdadynamodbstackLambdaFunctionServiceRole226532E1": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "testlambdadynamodbstackLambdaFunctionServiceRoleDefaultPolicyF537063F": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "testlambdadynamodbstackDynamoTable8138E93B",
                    "Arn"
                  ]
                },
                {
                  "Ref": "AWS::NoValue"
                }
              ]
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "testlambdadynamodbstackLambdaFunctionServiceRoleDefaultPolicyF537063F",
        "Roles": [
          {
            "Ref": "testlambdadynamodbstackLambdaFunctionServiceRole226532E1"
          }
        ]
      }
    },
    "testlambdadynamodbstackLambdaFunction5DDB3E8D": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters78e4557b6af07a75c658312f6045a0bb11a8ed0ed528857b3c0687477658ec9bS3Bucket8D2850DA"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters78e4557b6af07a75c658312f6045a0bb11a8ed0ed528857b3c0687477658ec9bS3VersionKeyB5B23FDF"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters78e4557b6af07a75c658312f6045a0bb11a8ed0ed528857b3c0687477658ec9bS3VersionKeyB5B23FDF"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "testlambdadynamodbstackLambdaFunctionServiceRole226532E1",
            "Arn"
          ]
        },
        "Runtime": "nodejs10.x",
        "Environment": {
          "Variables": {
            "LOGLEVEL": "INFO",
            "DDB_TABLE_NAME": {
              "Ref": "testlambdadynamodbstackDynamoTable8138E93B"
            }
          }
        }
      },
      "DependsOn": [
        "testlambdadynamodbstackLambdaFunctionServiceRoleDefaultPolicyF537063F",
        "testlambdadynamodbstackLambdaFunctionServiceRole226532E1"
      ]
    },
    "testlambdadynamodbstackLambdaFunctionLogRetention4724F084": {
      "Type": "Custom::LogRetention",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn"
          ]
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "testlambdadynamodbstackLambdaFunction5DDB3E8D"
              }
            ]
          ]
        },
        "RetentionInDays": 90
      }
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy"
              ],
              "Effect": "Allow",
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB"
          }
        ]
      }
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParametersb23e818e172b6771c5f4fde08381b8d0904b5760fcfcceee68523a83fff99c42S3Bucket773E39E4"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParametersb23e818e172b6771c5f4fde08381b8d0904b5760fcfcceee68523a83fff99c42S3VersionKey96D611D4"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParametersb23e818e172b6771c5f4fde08381b8d0904b5760fcfcceee68523a83fff99c42S3VersionKey96D611D4"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn"
          ]
        },
        "Runtime": "nodejs10.x"
      },
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB"
      ]
    }
  },
  "Parameters": {
    "AssetParameters78e4557b6af07a75c658312f6045a0bb11a8ed0ed528857b3c0687477658ec9bS3Bucket8D2850DA": {
      "Type": "String",
      "Description": "S3 bucket for asset \"78e4557b6af07a75c658312f6045a0bb11a8ed0ed528857b3c0687477658ec9b\""
    },
    "AssetParameters78e4557b6af07a75c658312f6045a0bb11a8ed0ed528857b3c0687477658ec9bS3VersionKeyB5B23FDF": {
      "Type": "String",
      "Description": "S3 key for asset version \"78e4557b6af07a75c658312f6045a0bb11a8ed0ed528857b3c0687477658ec9b\""
    },
    "AssetParameters78e4557b6af07a75c658312f6045a0bb11a8ed0ed528857b3c0687477658ec9bArtifactHashBE359D28": {
      "Type": "String",
      "Description": "Artifact hash for asset \"78e4557b6af07a75c658312f6045a0bb11a8ed0ed528857b3c0687477658ec9b\""
    },
    "AssetParametersb23e818e172b6771c5f4fde08381b8d0904b5760fcfcceee68523a83fff99c42S3Bucket773E39E4": {
      "Type": "String",
      "Description": "S3 bucket for asset \"b23e818e172b6771c5f4fde08381b8d0904b5760fcfcceee68523a83fff99c42\""
    },
    "AssetParametersb23e818e172b6771c5f4fde08381b8d0904b5760fcfcceee68523a83fff99c42S3VersionKey96D611D4": {
      "Type": "String",
      "Description": "S3 key for asset version \"b23e818e172b6771c5f4fde08381b8d0904b5760fcfcceee68523a83fff99c42\""
    },
    "AssetParametersb23e818e172b6771c5f4fde08381b8d0904b5760fcfcceee68523a83fff99c42ArtifactHashD6C3B3DD": {
      "Type": "String",
      "Description": "Artifact hash for asset \"b23e818e172b6771c5f4fde08381b8d0904b5760fcfcceee68523a83fff99c42\""
    }
  }
}